pkg/scheduler/queue/tree_queue_algo_querier_worker_queue_priority.go

```diff
@@ -4,74 +4,54 @@ package queue
+// During any period without change to the nodeOrder, the same worker ID consistently starts at the same queue node,
+// but moves on to other nodes if it cannot dequeue a request from the subtree of its first prioritized queue node.
+// Continuing to search through other query-component nodes and their subtrees minimizes idle querier-worker capacity.
 //
-//  3. The tree traversal algorithm recurs down to select child queue nodes of the "ingester" node,
-//     where each child queue node is a non-empty tenant-specific queue of ingester-only queries.
-//     The tenant-querier-shuffle-shard queue algorithm checks each tenant node for if it is sharded to this querier.
+// A querier-worker can process queries for nodes it has not prioritized when this QueuingAlgorithm is applied at the
+// highest layer of the tree and the tenant-querier-shuffle-shard QueuingAlgorithm applied at the second layer of the
+// tree. If shuffle-sharding is enabled, a querier-worker that prioritizes ingester-only queries may not find
+// ingester-only queries for any tenant it is assigned to, and move on to the next query component subtree. E.g.:
 //
-//  4. (a) The first tenant queue node found which is sharded to this querier will be dequeued from, and we are done.
+//  1. This algorithm has nodeOrder: ["ingester", "store-gateway", "ingester-and-store-gateway", "unknown"].
 //
-//  4. (b) Otherwise, if none of those tenants are sharded to this querier, the tree traversal algorithm will return
-//     back up to the parent level and ask the QuerierWorkerQueuePriorityAlgo to select its next node.
-//     We continue to step 5.
+//  2. A querier-worker with workerID 0 requests to dequeue; it prioritizes the "ingester" queue node.
 //
-//  5. The QuerierWorkerQueuePriorityAlgo will select the next node in the nodeOrder, "store-gateway".
-//     We return to step 3 and continue through steps 3, 4b, and 5, until we reach step 4a and exit.
+//  3. The dequeue operation attempts to dequeue first from the child nodes of the "ingester" node,
+//     where each child node is a tenant-specific queue of ingester-only queries. The tenantQuerierAssignments
+//     QueuingAlgorithm checks if any of its tenant queue nodes is sharded to this querier, and finds none.
 //
-// This process of continuing to search for requests to dequeue helps prevent querier-worker capacity from sitting idle
-// when there are no requests to dequeue for the query component node that the querier-worker was originally mapped to.
+//  4. The dequeue operation walks back up to the QuerierWorkerQueuePriorityAlgo level, not having dequeued anything.
+//     The QuerierWorkerQueuePriorityAlgo moves on and selects the next query-component node in the nodeOrder,
+//     and recurs again to search that next subtree for tenant queue nodes sharded to this querier, from step 3, etc.,
+//     until a dequeue-able tenant queue node is found, or every query component node subtree has been exhausted.
```

Comment:
Is it possible for a querier to find _no_ appropriate tenant node to dequeue from in the entire tree? This seems like it would be occasion for a reshuffle.

Reply:
(I'm not saying we should reshuffle in this case; just that this case:
> or every query component node subtree has been exhausted.

seems impossible to reach

Reply:
as far as the "could a querier not find a tenant node", this is not something we are changing, the querier could always possibly not find a tenant node but the only way I can think that it would actually occur would be some absurd shuffle shard config.

I think it would have to have num queriers > num tenants, and every tenant would have to have very few max queriers (like 1) so that some queriers would actually get shuffled to zero tenants.
I don't know if our shuffle shard logic guards against this at all but we dont use querier shuffle shard that much or that restrictively.

In that case a reshuffle wouldn't solve anyway.